Refactored API and new implementation

cquiroz opened the pull request:
This PR introduces a new faster algorithm to calculate the expt time and counts for a requested S/N

Aditionally I'v wanted to refactor the API for a while and while the old one is still present I'm switicing to a new one with a better response type and more fine grained error results

Feedback is welcome

cquiroz commented on modules/service/src/main/resources/graphql/itc.graphql:
```diff
@@ -4637,6 +4642,11 @@ type Query {
     input: SpectroscopyModeInput!
   ): [SpectroscopyResult!]!
 
+  spectroscopyExposureTime(
```
I'm considering whether to have a single `exposureTime` method that will take either a spectroscopy or an imaging mode input, WDYT?

toddburnside replied:
If it simplifies the code for you, it seems fine. But, if it doesn't, I'm not sure I see a reason to combine them. The client should know what they have.

rpiaggio commented on modules/service/src/main/scala/lucuma/itc/service/ItcMappings.scala:
```diff
@@ -72,16 +72,16 @@ object ItcMapping extends ItcCacheOrRemote with Version with GracklePartials {
           Schema(src.mkString).right.get
         }.liftTo[F]
       }
+    // .attemptTap(e => Logger[F].error(e.toString))
```
Commented debug code

rpiaggio commented on modules/service/src/main/resources/graphql/itc.graphql:
```diff
@@ -4830,6 +4840,89 @@ type SpectroscopyResult {
+}
+
+# ITC error result
+type ExposureTimeSourceTooBright {
```
`implements ExposureTimeResult` ?

cquiroz replied:
Good catch, that's right
t

rpiaggio commented on modules/service/src/main/resources/graphql/itc.graphql:
```diff
@@ -4830,6 +4840,89 @@ type SpectroscopyResult {
+}
+
+# ITC error result
+type ExposureTimeError {
```
`implements ExposureTimeResult` ?

rpiaggio commented on modules/service/src/main/resources/graphql/itc.graphql:
```diff
@@ -4830,6 +4840,89 @@ type SpectroscopyResult {
+# Possible exposure time itc result
+interface ExposureTimeResult {
+  # Result type
+  resultType: ExposureTimeResultType!
```
Are the subtypes and the enum both necessary? Clients can query what of the subtypes is returned by asking for `__typename`.

cquiroz replied:
ah you can tell when people know their stuff 😄

swalker2m reviewed (commented):
This all seems good to me.  Will the `itc-client` be updated to use the new API as part of this?

cquiroz commented on modules/itc/src/main/scala/lucuma/itc/legacy.scala:
```diff
@@ -0,0 +1,68 @@
+
+import scala.concurrent.duration.FiniteDuration
+
+case class LegacyItcResult(mode: ObservingMode, itc: LegacyExposureCalculationResult)
```
Yes the old api will be available for a little while

cquiroz commented on modules/service/src/main/resources/graphql/itc.graphql:
```diff
@@ -4830,6 +4840,89 @@ type SpectroscopyResult {
+  exposures: PosInt!
+
+  # Signal/Noise ratio
+  signalToNoise: PosBigDecimal!
```
Good point, I'll swich

cquiroz commented on modules/service/src/main/resources/graphql/itc.graphql:
```diff
@@ -4637,6 +4642,11 @@ type Query {
 
+  spectroscopyExposureTime(
+    # Query multiple modes in spectroscopy mode
+    input: SpectroscopyModeInput!
```
After some discussions with users, we may use just a single mode. This will simplify the api and implementation

cquiroz replied:
Do we normally name the input parameter `input`?

swalker2m replied:
> Do we normally name the input parameter input?

In the ODB schema for mutations we have a single input which we always call `input`.  The queries have various parameters.  What you did seems fine to me though.

cquiroz commented:
This could be reviewed again, The schema was changed as suggested in the comments.
As a result I removed the old API thus part of the work will be to update the clients (odb and explore)
Note that ITCCllient has been updated to match the new api

swalker2m commented on modules/client/shared/src/main/scala/lucuma/itc/client/ItcResult.scala:
```diff
@@ -16,65 +16,58 @@ import lucuma.core.util.TimeSpan
+      c.downField("__typename").as[String].flatMap {
+        case "ExposureEstimate" => c.as[Success].widen[ItcResult]
+        case "SourceTooBright"  => c.as[SourceTooBright].widen[ItcResult]
+        case "CalculationError" => c.as[Error].widen[ItcResult]
```
Should we make the Scala type names match the Schema type names?  `Success` => `ExposureTimeEstimate`, `Error` => `CalculationError`?

swalker2m commented on modules/client/shared/src/main/scala/lucuma/itc/client/SpectroscopyQuery.scala:
```diff
@@ -13,29 +13,26 @@ import io.circe.JsonObject
+            }
+            ... on CalculationError {
+              msg
             }
```
Don't we need to handle `SourceTooBright` as well?

cquiroz replied:
We do

swalker2m commented on modules/itc/src/main/scala/lucuma/itc/ExposureTime.scala:
```diff
@@ -0,0 +1,51 @@
+
+import scala.concurrent.duration.FiniteDuration
+
+sealed trait ExposureTimeResult extends Product with Serializable
```
Maybe a `sealed trait ExposureTimeError extends ExposureTimeResult` to group `CalculationError` and `SourceTooBright`. Thinking about it a bit more, I think it might help future proof the schema to have this in the schema as well.  Something like

```
interface ExposureTimeError {
  message: String!
}

union ExposureTimeResult = ExposureEstimate | SourceTooBright | CalculationError

type ExposureEstimate {
   # Exposure time duration
   exposureTime: Duration!

   # Exposures
   exposures: PosInt!

   # Signal/Noise ratio
   signalToNoise: SignalToNoise!
 }

 # ITC error when the source is too bright
 type SourceTooBright implements ExposureTimeError {
   message: String!

   # Half-well timme
   halfWellTime: BigDecimal!
 }

 # ITC generic error result
 type CalculationError implements ExposureTimeError {
  message: String!

   # Messages, the itc engine sometimes returns multiple error lines
   details: [String!]!
 }
```

That way a query that only cares about success vs. failure could "match" `... on ExposureTimeError` and not the individual cases.  Or you can add that and not worry about new error cases coming up in the future.

cquiroz replied:
Sounds very good, Would the union then be 
`union ExposureTimeResult = ExposureEstimate | ExposureTimeError`?

swalker2m replied:
Hmm, I don't know.  I think the `union` should still list the individual result cases. If you don't, I'm not sure you can then still match `... on CalculationError` specifically if you wanted to.

cquiroz commented:
I updated again the schema with three changes:

* Implemented the suggestions from @swalker2m about the error types
* Rename `ExposureTime` to `IntegrationTime` which is a more accurate term
* Use `TimeSpan` to represent the `exposureTime` values

swalker2m commented on modules/service/src/main/resources/graphql/itc.graphql:
```diff
@@ -4576,39 +4576,53 @@ scalar NonNegBigDecimal
-  # Signal/Noise ratio
-  signalToNoise: SignalToNoise!
+  spectroscopySignalToNoiseBeta(
+    input: SpectroscopySNInput!
```
A tiny, tiny, wee little detail but it might be good if it were `SignalToNoise` everywhere in the schema (it's only here that it is `SN`)?  We could also make the inputs match the query fields?

```
spectroscopyIntegrationTime(
  input: SpectroscopyIntegrationTimeInput!
): ExposureTimeCalculationResult!
```
 and
```
spectroscopySignalToNoise(
  input: SpectroscopySignalToNoiseInput!
): SignalToNoiseResult!
```
and
```
spectroscopyGraph(
  input: SpectroscopyGraphInput!
): SpectroscopyGraphResult!
```

When you have `imagingIntegrationTime` will the result also be an `ExposureTimeCalculationResult` (i.e., not specific to spectroscopy vs. imaging)?

cquiroz replied:
Good idea, I think the return for imaging will be the same for imaging and spectroscopy, the return type takes a mode which is a union of `ImagindMode | SpectroscopyMode`

swalker2m reviewed (approved):
Looks excellent! :+1:

swalker2m commented on modules/client/shared/src/main/scala/lucuma/itc/client/ItcResult.scala:
```diff
@@ -18,119 +18,51 @@ import java.math.MathContext
+  signalToNoise: SignalToNoise
+) extends ItcResult {
+
+  def stepSignalToNoise: Option[SignalToNoise] =
```
Hmm, should `exposures` be `PosInt`?  We'll get a divide by 0 error otherwise if 0.  Or maybe we return `None` if exposures is 0?

Maybe we should add this to `SignalToNoise` while we're at it?  In an `extension` I mean.  What do you think?

```
def stepSignalToNoise(exposures: NonNegInt): Option[SignalToNoise]
```

cquiroz replied:
Sure, exposures has to be 1 or higher right. I'm checking the SN 0 case and returning an error if that happens

cquiroz replied:
I realized we can reuse `IntegrationTime` on the server and client where the only difference is we use `PosInt` in one place

swalker2m replied:
> Sure, exposures has to be 1 or higher

Won't this throw an exception if it is called with 0 exposures?

cquiroz replied:
If it is a PosInt it can't be zero right?

cquiroz replied:
Yes, actually we can unify the model for client and server and use PosInt

swalker2m reviewed (commented):
I think the API seems simpler this way.

swalker2m commented:
Looks great!
